infra/token-transfer-server/src/lib/transfer.js

```diff
@@ -190,47 +195,81 @@ async function executeTransfer(transfer, transferTaskId) {
     txHash = await token.credit(transfer.toAddress, naturalAmount)
   } catch (error) {
     logger.error('Error crediting tokens', error.message)
+
     await updateTransferStatus(
       user,
       transfer,
       enums.TransferStatuses.Failed,
       TRANSFER_FAILED,
       error.message
     )
-    return { txHash: null, txStatus: enums.TransferStatuses.Failed }
+    return false
   }
 
+  logger.info(`Transfer ${transfer.id} processed with hash ${txHash}`)
+
   await transfer.update({
     status: enums.TransferStatuses.WaitingConfirmation,
     fromAddress: supplier.toLowerCase(),
     txHash
   })
 
+  return txHash
+}
+
+/**
+ * Sends a blockchain transaction to transfer tokens.
+ * @param {Transfer} transfer: DB model Transfer object
+ * @returns {Promise<>}
+ */
+async function checkBlockConfirmation(transfer) {
+  // Setup token library
+  const token = new Token(networkId)
+
   // Wait for the transaction to get confirmed.
-  const { status } = await token.waitForTxConfirmation(txHash, {
-    numBlocks: NumBlockConfirmation,
-    timeoutSec: ConfirmationTimeoutSec
+  const result = await token.txIsConfirmed(transfer.txHash, {
+    numBlocks: NumBlockConfirmation
   })
 
   let transferStatus, eventAction, failureReason
-  switch (status) {
-    case 'confirmed':
-      transferStatus = enums.TransferStatuses.Success
-      eventAction = TRANSFER_DONE
-      break
-    case 'failed':
-      transferStatus = enums.TransferStatuses.Failed
-      eventAction = TRANSFER_FAILED
-      break
-    case 'timeout':
+  if (!result) {
+    // Check if too long elapsed
+    const timedOut =
+      moment.utc(transfer.updatedAt).diff(moment.utc(), 'seconds') >
+      ConfirmationTimeoutSec
+    if (timedOut) {
       transferStatus = enums.TransferStatuses.Failed
       eventAction = TRANSFER_FAILED
       failureReason = 'Confirmation timeout'
```

Comment:
In case of a confirmation timeout, the tx may still get executed later on.
This may happen for ex if there is a surge in gas price right after the tx was submitted, causing the tx to not get mined for a while... If we consider the tx as failed the user may be able to double spend...
I wonder if it would be safer to halt all transfers until we get a chance to manually look at what happened?

Reply:
What about we just remove the timeout? In the case you described it'd be good if the script would eventually pick it up, even if it took a long time. And if there is a transaction that is still waiting for confirmation nothing else will process anyway so its effectively the same as halting all transfers.

Reply:
Yes I like your idea of removing the timeout. I'd vote for that too.